File: src/main/java/com/dude/dms/backend/brain/parsing/RuleValidator.java
package com.dude.dms.backend.brain.parsing;

import com.dude.dms.backend.data.docs.Doc;
import com.dude.dms.backend.data.rules.Rule;
import com.dude.dms.backend.data.tags.Tag;
import com.dude.dms.backend.service.DocService;
import com.dude.dms.backend.service.TagService;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public abstract class RuleValidator<T extends Rule> {

    protected final TagService tagService;

    protected final DocService docService;

    public abstract Set<Tag> getTagsForRule(String rawText, T rule);

    public abstract Set<Tag> getTags(String rawText);

    protected RuleValidator(TagService tagService, DocService docService) {
        this.tagService = tagService;
        this.docService = docService;
    }

    public Map<Doc, Set<Tag>> runRuleForAll(T rule) {
        Map<Doc, Set<Tag>> result = new HashMap<>();
        List<Doc> docs = docService.findAll();
        for (Doc doc : docs) {
            Set<Tag> tags = getTagsForRule(doc.getRawText(), rule);
            tags.removeAll(doc.getTags());
            if (!tags.isEmpty()) {
                result.put(doc, tags);
            }
        }
        return result;
    }
}

File: src/main/java/com/dude/dms/backend/brain/parsing/RuleValidator.kt
package com.dude.dms.backend.brain.parsing

import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.backend.data.rules.Rule
import com.dude.dms.backend.service.DocService
import com.dude.dms.backend.service.TagService
import java.util.*

abstract class RuleValidator<T : Rule>(protected val tagService: TagService, protected val docService: DocService) {

    abstract fun getTagsForRule(rawText: String?, rule: T): Set<Tag>

    abstract fun getTags(rawText: String?): Set<Tag?>?

    fun runRuleForAll(rule: T): Map<Doc, Set<Tag>> {
        val result: MutableMap<Doc, Set<Tag>> = HashMap()
        val docs = docService.findAll()
        for (doc in docs) {
            val tags = getTagsForRule(doc.rawText, rule).toMutableSet().apply { removeAll(doc.tags) }
            if (tags.isNotEmpty()) {
                result[doc] = tags
            }
        }
        return result
    }
}